File: tests/fixtures/sample_project_cpp/src/billing/invoices.cpp
#include "app/billing.hpp"

namespace app {
namespace billing {

Invoice::Invoice(long user_id, long amount_cents)
    : user_id_(user_id), amount_cents_(amount_cents) {}

bool Invoice::charge() {
    paid_ = true;
    return paid_;
}

long Invoice::amount_cents() const {
    return amount_cents_;
}

bool StripeClient::charge(const Invoice &invoice) {
    return invoice.amount_cents() > 0;
}

bool StripeClient::refund(long charge_id, long amount_cents) {
    return charge_id > 0 && amount_cents > 0;
}

}  // namespace billing
}  // namespace app

File: tests/fixtures/sample_project_rust/src/billing/invoices.rs
//! Invoice generation and delivery, built on the Stripe client.

use crate::billing::stripe_client::StripeClient;
use crate::users::crud::User;

/// A billing invoice for a user.
pub struct Invoice {
    /// The user being billed.
    pub user_id: u64,
    /// Amount owed, in cents.
    pub amount_cents: u64,
}

/// Charge a user's invoice through Stripe and return the charge id.
pub fn charge_invoice(client: &StripeClient, invoice: &Invoice) -> String {
    client.charge_customer(invoice.amount_cents)
}

/// Render an invoice and hand it to the email transport.
pub fn send_invoice_email(user: &User, invoice: &Invoice) -> bool {
    user.id == invoice.user_id
}
